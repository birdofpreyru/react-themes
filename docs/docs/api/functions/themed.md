# themed()
[themed()]: themed
```tsx
import themed from '@dr.pogodin/react-themes';

// Recommended signature:
function themed<ComponentProps extends ThemeableComponentProps>(
  component: ComponentType<ComponentProps>,
  componentName: string,
  defaultTheme: ComponentProps['theme'],
  options?: ThemedOptions<ComponentProps>,
): ThemedComponent<ComponentProps>;
```

<details>
<summary>Full & decorator signatures of [themed()]</summary>

The full signature of [themed()] function is the following
(essentially, it allows to skip `defaultTheme` argument altogether, and
to provide `options` in its position). It is provided mostly for backward
compatibility, as in practice `defaultTheme` argument is always used.
```tsx
function themed<ComponentProps extends ThemeableComponentProps>(
  component: ComponentType<ComponentProps>,
  componentName: string,

  defaultThemeOrOptions?: ComponentProps['theme']
    | ThemedOptions<ComponentProps>,

  options?: ThemedOptions<ComponentProps>,
): ThemedComponent<ComponentProps>;
```
Also for backward compatibility, this variant of [themed()] function exists,
and allows to use it as a (legacy?) decorator. It is not recommended,
as [Decorators](https://github.com/tc39/proposal-decorators) in JavaScript
have a difficult history, with different variants implemented by different
tools ([Babel](https://babeljs.io/docs/babel-plugin-proposal-decorators)),
and language flavours
([TypeScript](https://www.typescriptlang.org/docs/handbook/decorators.html))
at different times, and because decorators do not provide any unqiue
possibilities that can't be achived without them. If ever they finally become
a standard part of JS/TS eco-system, most probably their support by this library
will have to be revised.
```tsx
function themed<ComponentProps extends ThemeableComponentProps>(
  componentName: string,

  defaultThemeOrOptions?: ComponentProps['theme']
    | ThemedOptions<ComponentProps>,

  options?: ThemedOptions<ComponentProps>,
): ThemedComponentFactory<ComponentProps>;
```
</details>


Registers a themed component under the given name, and with an optional default
theme.

[themed()] can be used as a decorator, or in the following (recommended) way:
```jsx
import themed from '@dr.pogodin/react-themes';
import defaultTheme from './default.scss';

function Component() { ... }

export themed(Component, 'ThemedComponent', defaultTheme);
```

When rendered, your component will receive the composed theme via its
`theme` prop. You will just need to pass the values from `theme` into
the `className` attributes of your component elements, as shown in
the [Getting Started](/docs/tutorial/getting-started#basic-themed-component)
tutorial.

## Generic Parameter
- `ComponentProps` extends [ThemeableComponentProps] &mdash; this type describes
  the props accepted by the original component, wrapped by [themed()] function.

## Arguments
- `component` &mdash; **ComponentType\[ComponentProps\]** &mdash; The base
  component to be themed, its props must compty with the [ThemeableComponentProps]
  interface.

- `componentName` &mdash; **string** &mdash; Themed component name, which should
  be used to provide its context theme via [ThemeProvider].

- `defaultTheme` &mdash; **ComponentProps\['theme'\]** &mdash; Default theme,
  in the form of theme key to CSS classname mapping.

**Optional**

- `options` &mdash; [ThemedOptions] &mdash; Additional options accepted by
  the function.

## Returns

The recommended signature of this function returns themed component:
- [ThemedComponent]**\<ComponentProps\>**

The decorator variant returns instead the component factory, which should be
called right away with a single argument &mdash; the base component to wrap:
- [ThemedComponentFactory]**\<ComponentProps\>**

[ThemeableComponentProps]: /docs/api/types/themeablecomponentprops
[ThemedComponent]: /docs/api/types/themedcomponent
[ThemedComponentFactory]: /docs/api/types/themedcomponentfactory
[ThemedOptions]: /docs/api/types/themedoptions
[ThemeProvider]: /docs/api/components/themeprovider
