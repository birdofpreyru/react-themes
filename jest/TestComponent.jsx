import PT from 'prop-types';

export default function Component({ children, theme }) {
  return (
    <div className={theme.container}>
      <div className={theme.content}>
        { children }
      </div>
    </div>
  );
}

Component.propTypes = {
  children: PT.node,
  theme: PT.shape({
    container: PT.string,
    content: PT.string,
  }).isRequired,
};

Component.defaultProps = {
  children: null,
};
