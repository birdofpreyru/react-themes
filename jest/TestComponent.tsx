import type { FunctionComponent } from 'react';

import type { Theme } from '../src';

export type ComponentTheme = Theme<'container' | 'content'>;

export type ComponentProps = {
  children?: React.ReactNode;
  theme: ComponentTheme;
  goodKey?: string;
};

const Component: FunctionComponent<ComponentProps> = ({
  children,
  goodKey = '',
  theme,
}) => (
  <div className={theme.container}>
    <div className={theme.content}>
      { children }
      { goodKey }
    </div>
  </div>
);

export default Component;
