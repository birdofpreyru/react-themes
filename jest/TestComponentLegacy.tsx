// This is a component accepting legacy themes (Themr library).

export type ComponentThemeT = {
  container?: string;
  content?: string;
};

type ComponentPropsT = {
  children?: React.ReactNode;
  theme: ComponentThemeT;
};

export default function Component({ children, theme }: ComponentPropsT) {
  return (
    <div className={theme.container}>
      <div className={theme.content}>
        { children }
      </div>
    </div>
  );
}

Component.defaultProps = {
  children: null,
};
