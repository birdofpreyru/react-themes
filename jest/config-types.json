{
  "rootDir": "..",
  "runner": "jest-runner-tsd",
  "testMatch": ["**/__tests__/ts-types/**/*.(ts|tsx)"]
}
