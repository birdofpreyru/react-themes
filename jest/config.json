{
  "collectCoverage": true,
  "collectCoverageFrom": ["src/**/*.{js,jsx,ts,tsx}"],
  "coverageDirectory": "__coverage__",
  "rootDir": "..",
  "testPathIgnorePatterns": [
    "/__tests__/ts-types/"
  ],
  "transform": {
    "\\.(j|t)sx?$": "babel-jest"
  }
}
