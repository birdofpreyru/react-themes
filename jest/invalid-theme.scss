/* Example of invalid theme. */

.container {
  color: red;
}

.unexpectedField {
  color: green;
}
