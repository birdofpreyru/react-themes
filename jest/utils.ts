/**
 * A few utils for Jest testing.
 */

/* global expect */

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
function render(component: React.ReactElement) {
  return Renderer.create(component).toJSON();
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component: React.ReactElement) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export default null;
