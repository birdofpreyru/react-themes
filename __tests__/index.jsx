import themed, { COMPOSE, PRIORITY, ThemeProvider } from '../src';

import { snapshot } from '../jest/utils';
import TestComponent from '../jest/TestComponent';

import themeA from '../jest/theme-a.scss';
import themeB from '../jest/theme-b.scss';
import themeC from '../jest/theme-c.scss';
import themeD from '../jest/theme-d.scss';

describe('01 - No default theme.', () => {
  describe('01 - No options at registration.', () => {
    const Themed = themed('Themed')(TestComponent);

    test('01 - No extra themes', () => {
      snapshot(<Themed />);
      snapshot(<Themed>Child</Themed>);
    });

    test('02 - Ad hoc theme', () => {
      snapshot(<Themed theme={themeA}>Child</Themed>);
    });

    test('03 - Context theme', () => {
      snapshot((
        <ThemeProvider themes={{}}>
          <Themed />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed composeAdhocTheme={COMPOSE.SWAP} />
        </ThemeProvider>
      ));
    });

    test('04 - Context + ad hoc', () => {
      snapshot((
        <ThemeProvider themes={{}}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeA }}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeA }}>
          <ThemeProvider themes={{ Themed: themeB }}>
            <Themed theme={themeC} />
          </ThemeProvider>
        </ThemeProvider>
      ));
    });

    test('05 - Context + ad hoc + composition modes', () => {
      snapshot((
        <ThemeProvider themes={{ Themed: themeA }}>
          <Themed composeAdhocTheme={COMPOSE.SOFT} theme={themeB} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeA }}>
          <Themed composeAdhocTheme={COMPOSE.SWAP} theme={themeB} />
        </ThemeProvider>
      ));

      expect(
        () => snapshot((
          <ThemeProvider themes={{ Themed: themeA }}>
            <Themed composeAdhocTheme="deeply" theme={themeB} />
          </ThemeProvider>
        )),
      ).toThrowErrorMatchingSnapshot();
    });

    test(
      '06 - Nested ThemeProvider does not shadow parent themes it does not override',
      () => {
        const Themed2 = themed('Themed2')(TestComponent);
        snapshot((
          <ThemeProvider themes={{ Themed: themeA, Themed2: themeA }}>
            <ThemeProvider themes={{ Themed2: themeB }}>
              <Themed />
              <Themed2 />
            </ThemeProvider>
          </ThemeProvider>
        ));
      },
    );
  });
});

describe('02 - With default theme', () => {
  describe('01 - No options at registration.', () => {
    const Themed = themed('Themed', themeA)(TestComponent);

    test('01 - No extra themes', () => {
      snapshot(<Themed />);
      snapshot(<Themed>Child</Themed>);
    });

    test('02 - Ad hoc theme', () => {
      snapshot(<Themed theme={themeB}>Child</Themed>);
    });

    test('03 - Context theme', () => {
      snapshot((
        <ThemeProvider themes={{}}>
          <Themed />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed composeContextTheme={COMPOSE.SWAP} />
        </ThemeProvider>
      ));
    });

    test('04 - Context + ad hoc', () => {
      snapshot((
        <ThemeProvider themes={{}}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeC }}>
          <Themed theme={themeD} />
        </ThemeProvider>
      ));
    });

    test('05 - Context + ad hoc + composition modes', () => {
      snapshot((
        <ThemeProvider themes={{ Themed: themeC }}>
          <Themed
            themePriority={PRIORITY.ADHOC_DEFAULT_CONTEXT}
            theme={themeD}
          />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeC }}>
          <Themed
            composeContextTheme={COMPOSE.SWAP}
            themePriority={PRIORITY.ADHOC_DEFAULT_CONTEXT}
            theme={themeD}
          />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeC }}>
          <Themed
            composeContextTheme={COMPOSE.SWAP}
            theme={themeD}
          />
        </ThemeProvider>
      ));
    });

    test('06 - Theme props mapping', () => {
      let args;
      snapshot((
        <Themed
          mapThemeProps={(props, theme) => {
            args = { props, theme };
            return {
              theme: {},
            };
          }}
        />
      ));
      expect(args).toMatchSnapshot();
    });
  });

  describe('02 - Options at registration', () => {
    test('01 - composeContextTheme', () => {
      const Themed = themed('Themed', themeA, {
        composeContextTheme: COMPOSE.SWAP,
      })(TestComponent);
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed />
        </ThemeProvider>
      ));
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed composeContextTheme={COMPOSE.SOFT} />
        </ThemeProvider>
      ));
    });

    test('02 - composeAdhocTheme', () => {
      const Themed = themed('Themed', themeA, {
        composeAdhocTheme: COMPOSE.SWAP,
      })(TestComponent);
      snapshot((
        <Themed theme={themeB} />
      ));
      snapshot((
        <Themed theme={themeB} composeAdhocTheme={COMPOSE.SOFT} />
      ));
    });

    test('03 - themePriority', () => {
      const Themed = themed('Themed', themeA, {
        themePriority: PRIORITY.ADHOC_DEFAULT_CONTEXT,
      })(TestComponent);
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed theme={themeC} />
          <Themed
            theme={themeC}
            themePriority={PRIORITY.ADHOC_CONTEXT_DEFAULT}
          />
        </ThemeProvider>
      ));
    });

    test('04 - Theme props mapping', () => {
      let args;
      const Themed = themed('Themed', themeA, {
        mapThemeProps: (props, theme) => {
          args = { props, theme };
          return { theme: {} };
        },
      })(TestComponent);
      snapshot(<Themed />);
      expect(args).toMatchSnapshot();
      snapshot((
        <Themed
          mapThemeProps={(props, theme) => {
            args = { props, theme };
            return { theme };
          }}
        />
      ));
      expect(args).toMatchSnapshot();
    });

    test('05 - Specifity tags', () => {
      const Themed = themed('Themed', themeA, {
        adhocTag: 'absent.hoc',
        contextTag: 'absent',
      })(TestComponent);
      snapshot((
        <ThemeProvider themes={{ Themed: themeB }}>
          <Themed theme={themeC} />
        </ThemeProvider>
      ));
      expect(
        () => themed('Themed', themeA, {
          adhocTag: '.invalid',
          contextTag: 'absent',
        })(TestComponent),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
