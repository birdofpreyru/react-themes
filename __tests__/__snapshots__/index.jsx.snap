// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01 - No default theme. 01 - No options at registration. 01 - No extra themes 1`] = `
<div>
  <div />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 01 - No extra themes 2`] = `
<div>
  <div>
    Child
  </div>
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 02 - Ad hoc theme 1`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k"
  >
    Child
  </div>
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 03 - Context theme 1`] = `
<div>
  <div />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 03 - Context theme 2`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 03 - Context theme 3`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 04 - Context + ad hoc 1`] = `
<div>
  <div
    className="theme-c__content___xSnq1"
  />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 04 - Context + ad hoc 2`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div
    className="theme-c__content___xSnq1"
  />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 04 - Context + ad hoc 3`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k theme-c__ad___i7Ven theme-c__hoc___6A2dq theme-c__content___xSnq1"
  />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 04 - Context + ad hoc 4`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div
    className="theme-c__content___xSnq1"
  />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 05 - Context + ad hoc + composition modes 1`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 05 - Context + ad hoc + composition modes 2`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`01 - No default theme. 01 - No options at registration. 05 - Context + ad hoc + composition modes 3`] = `"Invalid composition mode"`;

exports[`01 - No default theme. 01 - No options at registration. 06 - Nested ThemeProvider does not shadow parent themes it does not override 1`] = `
Array [
  <div
    className="theme-a__container___OHZan"
  >
    <div
      className="theme-a__content___IgY-k"
    />
  </div>,
  <div
    className="theme-b__container___mfrvK"
  >
    <div />
  </div>,
]
`;

exports[`02 - With default theme 01 - No options at registration. 01 - No extra themes 1`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 01 - No extra themes 2`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k"
  >
    Child
  </div>
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 02 - Ad hoc theme 1`] = `
<div
  className="theme-a__container___OHZan theme-b__ad___MToBJ theme-b__hoc___D-90l theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k"
  >
    Child
  </div>
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 03 - Context theme 1`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 03 - Context theme 2`] = `
<div
  className="theme-a__container___OHZan theme-b__context___623DE theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 03 - Context theme 3`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 04 - Context + ad hoc 1`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k theme-c__ad___i7Ven theme-c__hoc___6A2dq theme-c__content___xSnq1"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 04 - Context + ad hoc 2`] = `
<div
  className="theme-a__container___OHZan theme-b__context___623DE theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k theme-c__ad___i7Ven theme-c__hoc___6A2dq theme-c__content___xSnq1"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 04 - Context + ad hoc 3`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k theme-c__context___c253P theme-c__content___xSnq1 theme-d__ad___WibJk theme-d__hoc___ycFZY theme-d__content___xEs5l"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 05 - Context + ad hoc + composition modes 1`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-c__content___xSnq1 theme-a__context___YNfwL theme-a__content___IgY-k theme-d__ad___WibJk theme-d__hoc___ycFZY theme-d__content___xEs5l"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 05 - Context + ad hoc + composition modes 2`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k theme-d__ad___WibJk theme-d__hoc___ycFZY theme-d__content___xEs5l"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 05 - Context + ad hoc + composition modes 3`] = `
<div>
  <div
    className="theme-c__content___xSnq1 theme-d__ad___WibJk theme-d__hoc___ycFZY theme-d__content___xEs5l"
  />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 06 - Theme props mapping 1`] = `
<div>
  <div />
</div>
`;

exports[`02 - With default theme 01 - No options at registration. 06 - Theme props mapping 2`] = `
Object {
  "props": Object {
    "mapThemeProps": [Function],
    "ref": null,
  },
  "theme": Object {
    "ad": "theme-a__ad___yK+Km",
    "container": "theme-a__container___OHZan",
    "content": "theme-a__content___IgY-k",
    "context": "theme-a__context___YNfwL",
    "hoc": "theme-a__hoc___zE9Bs",
  },
}
`;

exports[`02 - With default theme 02 - Options at registration 01 - composeContextTheme 1`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 01 - composeContextTheme 2`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 02 - composeAdhocTheme 1`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 02 - composeAdhocTheme 2`] = `
<div
  className="theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 03 - themePriority 1`] = `
Array [
  <div
    className="theme-b__container___mfrvK theme-a__context___YNfwL theme-a__container___OHZan"
  >
    <div
      className="theme-a__content___IgY-k theme-c__ad___i7Ven theme-c__hoc___6A2dq theme-c__content___xSnq1"
    />
  </div>,
  <div
    className="theme-a__container___OHZan theme-b__context___623DE theme-b__container___mfrvK"
  >
    <div
      className="theme-a__content___IgY-k theme-c__ad___i7Ven theme-c__hoc___6A2dq theme-c__content___xSnq1"
    />
  </div>,
]
`;

exports[`02 - With default theme 02 - Options at registration 04 - Theme props mapping 1`] = `
<div>
  <div />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 04 - Theme props mapping 2`] = `
Object {
  "props": Object {
    "ref": null,
  },
  "theme": Object {
    "ad": "theme-a__ad___yK+Km",
    "container": "theme-a__container___OHZan",
    "content": "theme-a__content___IgY-k",
    "context": "theme-a__context___YNfwL",
    "hoc": "theme-a__hoc___zE9Bs",
  },
}
`;

exports[`02 - With default theme 02 - Options at registration 04 - Theme props mapping 3`] = `
<div
  className="theme-a__container___OHZan"
>
  <div
    className="theme-a__content___IgY-k"
  />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 04 - Theme props mapping 4`] = `
Object {
  "props": Object {
    "mapThemeProps": [Function],
    "ref": null,
  },
  "theme": Object {
    "ad": "theme-a__ad___yK+Km",
    "container": "theme-a__container___OHZan",
    "content": "theme-a__content___IgY-k",
    "context": "theme-a__context___YNfwL",
    "hoc": "theme-a__hoc___zE9Bs",
  },
}
`;

exports[`02 - With default theme 02 - Options at registration 05 - Specifity tags 1`] = `
<div
  className="theme-a__container___OHZan  theme-b__container___mfrvK"
>
  <div
    className="theme-a__content___IgY-k  theme-c__hoc___6A2dq theme-c__content___xSnq1"
  />
</div>
`;

exports[`02 - With default theme 02 - Options at registration 05 - Specifity tags 2`] = `"Invalid adhoc theme tag"`;
